File: municipality-dashboard/src/app/services/report-analytics.service.ts
import { environment } from 'src/environments/environment';
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

const BASE = environment.apiUrl + '/report';

export interface SummaryKPI {
  totalCitizens:      number;
  totalServices:      number;
  pendingRequests:    number;
  completedThisMonth: number;
  totalRevenue:       number;
  taxDueCount:        number;
}

export interface ChartPoint {
  label: string;  
  value: number;
}

@Injectable({ providedIn: 'root' })
export class ReportAnalyticsService {

  constructor(private http: HttpClient) {}

  // ── Summary KPIs 
  getSummary(): Observable<SummaryKPI> {
    return this.http.get<SummaryKPI>(`${BASE}/analytics/summary`);
  }

  // ── Citizen Report
  getCitizenReport(): Observable<any[]> {
    return this.http.get<any[]>(`${BASE}/citizens`);
  }
  getCitizensByWard(): Observable<any[]> {
    return this.http.get<any[]>(`${BASE}/citizens/by-ward`);
  }
  getCitizenGenderDistribution(): Observable<Record<string, number>> {
    return this.http.get<Record<string, number>>(`${BASE}/citizens/gender-distribution`);
  }

  // ── Service Report
  getServiceReport(): Observable<any[]> {
    return this.http.get<any[]>(`${BASE}/services`);
  }
  getServicesByType(): Observable<any[]> {
    return this.http.get<any[]>(`${BASE}/services/by-type`);
  }
  getServicesByStatus(): Observable<Record<string, number>> {
    return this.http.get<Record<string, number>>(`${BASE}/services/by-status`);
  }

  // ── Analytics 
  getMonthlyAnalytics(year: number): Observable<any[]> {
    const params = new HttpParams().set('year', year);
    return this.http.get<any[]>(`${BASE}/analytics/monthly`, { params });
  }
  getYearlyAnalytics(fromYear: number, toYear: number): Observable<any[]> {
    const params = new HttpParams().set('fromYear', fromYear).set('toYear', toYear);
    return this.http.get<any[]>(`${BASE}/analytics/yearly`, { params });
  }

  // ── Tax Collection 
  getTaxCollectionReport(): Observable<any[]> {
    return this.http.get<any[]>(`${BASE}/tax`);
  }
  getTaxByWard(): Observable<any[]> {
    return this.http.get<any[]>(`${BASE}/tax/by-ward`);
  }
  getMonthlyTaxCollection(year: number): Observable<any[]> {
    const params = new HttpParams().set('year', year);
    return this.http.get<any[]>(`${BASE}/tax/monthly`, { params });
  }
}

File: governance/src/main/java/com/mgt/service/ReportAnalyticsService.java
package com.mgt.service;

import java.util.*;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.mgt.dao.ReportAnalyticsDAO;

@Service
public class ReportAnalyticsService {

    @Autowired
    private ReportAnalyticsDAO reportAnalyticsDAO;

    //Citizen Report 
    public List<Map<String, Object>> getCitizenReport() {
        return reportAnalyticsDAO.fetchAllCitizens();
    }

    public List<Map<String, Object>> getCitizensByWard() {
        return reportAnalyticsDAO.countCitizensByWard();
    }

    public Map<String, Long> getGenderDistribution() {
        return reportAnalyticsDAO.getCitizenGenderDistribution();
    }

    // Service Report 
    public List<Map<String, Object>> getServiceReport() {
        return reportAnalyticsDAO.fetchAllServiceRequests();
    }

    public List<Map<String, Object>> getServicesByType() {
        return reportAnalyticsDAO.countServicesByType();
    }

    public Map<String, Long> getServicesByStatus() {
        return reportAnalyticsDAO.countServicesByStatus();
    }

    // Monthly / Yearly Analytics
    public List<Map<String, Object>> getMonthlyAnalytics(int year) {
        Map<Integer, Long> raw = reportAnalyticsDAO.countServicesByMonth(year);

        String[] monthNames = {"Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"};
        List<Map<String, Object>> result = new ArrayList<>();
        for (int m = 1; m <= 12; m++) {
            Map<String, Object> point = new LinkedHashMap<>();
            point.put("month", monthNames[m - 1]);
            point.put("count", raw.getOrDefault(m, 0L));
            result.add(point);
        }
        return result;
    }

    public List<Map<String, Object>> getYearlyAnalytics(int fromYear, int toYear) {
        Map<Integer, Long> raw = reportAnalyticsDAO.countServicesByYear(fromYear, toYear);
        List<Map<String, Object>> result = new ArrayList<>();
        for (int y = fromYear; y <= toYear; y++) {
            Map<String, Object> point = new LinkedHashMap<>();
            point.put("year", String.valueOf(y));
            point.put("count", raw.getOrDefault(y, 0L));
            result.add(point);
        }
        return result;
    }

    public Map<String, Object> getSummary() {
        Map<String, Object> summary = new LinkedHashMap<>();
        summary.put("totalCitizens",       reportAnalyticsDAO.countTotalCitizens());
        summary.put("totalServices",       reportAnalyticsDAO.countTotalServices());
        summary.put("pendingRequests",     reportAnalyticsDAO.countPendingRequests());
        summary.put("completedThisMonth",  reportAnalyticsDAO.countCompletedThisMonth());
        summary.put("totalRevenue",        reportAnalyticsDAO.sumTotalRevenue());
        summary.put("taxDueCount",         reportAnalyticsDAO.countTaxDue());
        return summary;
    }

    // Tax Collection Report 

    public List<Map<String, Object>> getTaxCollectionReport() {
        return reportAnalyticsDAO.fetchTaxPayments();
    }

    public List<Map<String, Object>> getTaxByWard() {
        return reportAnalyticsDAO.sumTaxByWard();
    }

    public List<Map<String, Object>> getMonthlyTaxCollection(int year) {
        Map<Integer, Double> raw = reportAnalyticsDAO.sumTaxByMonth(year);
        String[] monthNames = {"Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"};
        List<Map<String, Object>> result = new ArrayList<>();
        for (int m = 1; m <= 12; m++) {
            Map<String, Object> point = new LinkedHashMap<>();
            point.put("month", monthNames[m - 1]);
            point.put("amount", raw.getOrDefault(m, 0.0));
            result.add(point);
        }
        return result;
    }
}
